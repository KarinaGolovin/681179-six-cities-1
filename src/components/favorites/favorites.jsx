import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import PlaceCard from '../place-card/place-card.jsx';
import {connect} from 'react-redux';
import {getFavoriteOfferList} from '../../store/actions';
import {withProps} from 'recompose';
import {getFavoriteOffersByCities, getFavoritesGroupedWithCities} from '../../store/reducers/offers/selectors';

const PlaceCardWrapped = withProps({
  classes: {
    container: `favorites__card`,
    imageWrapper: `favorites__image-wrapper`,
    cardInfo: `favorites__card-info`
  }
})(PlaceCard);

export class Favorites extends PureComponent {
  componentDidMount() {
    const {loadFavorites = () => {}} = this.props;

    loadFavorites();
  }
  render() {
    const {favoriteList: list, favoritesByCity = [], onCityClick, onBookmarkClick} = this.props;

    return (
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className={`favorites ${!list.length ? `favorites--empty` : ``}`}>
            {
              !list.length ? <>
                <h1 className="visually-hidden">Favorites (empty)</h1>
                <div className="favorites__status-wrapper">
                  <b className="favorites__status">Nothing yet saved.</b>
                  <p className="favorites__status-description">Save properties to narrow down search or plan yor future
                  trips.</p>
                </div>
              </> : <>
                  <h1 className="favorites__title">Saved listing</h1>
                  <ul className="favorites__list">
                    {favoritesByCity.map(([city, cards]) => {
                      return <FavoriteLocation
                        city={city}
                        cards={cards}
                        key={city}
                        onCityClick={onCityClick}
                        onBookmarkClick={onBookmarkClick}
                      />;
                    })}
                  </ul>
                </>
            }
          </section>
        </div>
      </main>
    );
  }
}

Favorites.propTypes = {
  favoriteList: PropTypes.array,
  favoritesByCity: PropTypes.array,
  onCityClick: PropTypes.func,
  onBookmarkClick: PropTypes.func,
  loadFavorites: PropTypes.func
};

const FavoriteLocation = ({city, cards, onCityClick, onBookmarkClick}) => {
  return (
    <>
      <li className="favorites__locations-items">
        <div className="favorites__locations locations locations--current">
          <div className="locations__item">
            <a className="locations__item-link" href="#" onClick={onCityClick}>
              <span>{city}</span>
            </a>
          </div>
        </div>
        <div className="favorites__places">
          {cards.map((it) => {
            return (
              <PlaceCardWrapped
                title={it.title}
                type={it.type}
                price={it.price}
                previewImage={it.preview_image}
                rating={it.rating}
                isPremium={it.is_premium}
                isBookmarked={it.is_favorite}
                onBookmarkClick={() => {
                  onBookmarkClick({
                    hotelId: it.id,
                    status: it.is_favorite ? 0 : 1
                  });
                }}
                id={it.id}
                key={it.id}
                imageWidth={150}
                imageHeight={110}
              />
            );
          })}
        </div>
      </li>
    </>
  );
};

FavoriteLocation.propTypes = {
  cards: PropTypes.array,
  city: PropTypes.string,
  onCityClick: PropTypes.func,
  onBookmarkClick: PropTypes.func
};

const mapStateToProps = (state) => {
  return {
    favoriteList: getFavoriteOffersByCities(state),
    favoritesByCity: getFavoritesGroupedWithCities(state)
  };
};

const mapDispatchToProps = {
  loadFavorites: getFavoriteOfferList
};

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
