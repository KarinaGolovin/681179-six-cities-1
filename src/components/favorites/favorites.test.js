import React from 'react';
import renderer from 'react-test-renderer';
import {Favorites} from './favorites.jsx';

const mock = [
  {Test: [
    {
      title: `Title`,
      type: `Type`,
      price: 1,
      previewImage: `Img`,
      link: `Link`,
      onLinkClick: (() => {}),
    }
  ]}
];

it(`Favorites correctly renders`, () => {
  const component = renderer.create(
      <Favorites
        onCityClick={() => {}}
        favoriteList={mock}
      />
  ).toJSON();

  expect(component).toMatchSnapshot();
});
