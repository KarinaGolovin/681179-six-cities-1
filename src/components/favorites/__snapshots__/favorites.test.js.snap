// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites correctly renders 1`] = `
<main
  className="page__main page__main--favorites"
>
  <div
    className="page__favorites-container container"
  >
    <section
      className="favorites "
    >
      <h1
        className="favorites__title"
      >
        Saved listing
      </h1>
      <ul
        className="favorites__list"
      >
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
                onClick={[Function]}
              >
                <span>
                  Test
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="place-card favorites__card"
              id={1}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__image-wrapper favorites__image-wrapper"
              >
                <a
                  href="#"
                  onClick={[Function]}
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height={110}
                    width={150}
                  />
                </a>
              </div>
              <div
                className="place-card__info favorites__card-info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      0
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button  button"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="rating place-card__rating"
                >
                  <div
                    className="rating__stars place-card__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "0%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="#"
                  >
                    Test
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  Test
                </p>
              </div>
            </article>
          </div>
        </li>
      </ul>
    </section>
  </div>
</main>
`;
